modules/ui/src/main/scala/lucuma/ui/optics/ChangeAuditor.scala

```diff
@@ -0,0 +1,320 @@
+          case degrees :: minutes :: seconds :: Nil =>
+            isValidNDigitInt(degrees, 2, 90) && isValidNDigitInt(minutes, 2, 59) &&
+              isValidSeconds(seconds, 6)
+          case _                                    => false
+        }
+
+      if (isValid)
+        if (noSign == stripped) AuditResult.accept else AuditResult.newString(s"$sign$stripped")
+      else
+        AuditResult.reject
+    }
+
+  /**
+   * Build from an InputFormat instance.
+   */
+  def fromFormat[A](f: InputFormat[A]): ChangeAuditor[A] = ChangeAuditor { (s, _) =>
+    f.getOption(s).fold(AuditResult.reject)(_ => AuditResult.accept)
+  }
+
+  /**
+   * Build from a ValidFormatInput instance.
+   */
+  def fromValidFormatInput[A](v: ValidFormatInput[A]): ChangeAuditor[A] = ChangeAuditor { (s, _) =>
+    v.getValidated(s).fold(_ => AuditResult.reject, _ => AuditResult.accept)
+  }
+
+  private def fixIntString(str: String, cursorPos: Int): (String, String, Int) =
+    str match {
+      case ""   => ("0", str, 0)
+      case "-"  => ("-0", str, 0)
+      case "0-" => ("0", "-", -1)
+      case _    => stripZerosBeforeN(str, cursorPos)
+    }
+
+  private def fixDecimalString(
+    str:       String,
+    cursorPos: Int,
+    decimals:  Int
+  ): (String, String, Int) = {
+    val postStripped = stripZerosPastNPlaces(str, decimals)
+    postStripped match {
+      case ""   => ("0", postStripped, 0)
+      case "-"  => ("-0", postStripped, 0)
+      case "0-" => ("0", "-", -1)
+      case "."  => ("0.0", "0.", 1)
+      case _    =>
+        val dp = postStripped.indexOf(".")
+        val n  = if (dp < 0) cursorPos else math.min(dp, cursorPos)
+        stripZerosBeforeN(postStripped, n)
+    }
+  }
+
+  private def stripZerosPastNPlaces(str: String, n: Int): String = {
+    val regex = s"(.*\\.\\d{0,$n}[1-9]*)+0*".r
+    str match {
+      case regex(base) => base
+      case _           => str
+    }
+  }
+
+  private def stripZerosBeforeN(str: String, n: Int): (String, String, Int) = {
+    val (minus, newStr, newPos) =
+      if (str.startsWith("-")) ("-", str.substring(1), n - 1) else ("", str, n)
+    if (newPos > 0) {
+      // We actually only want to strip zeros if there is another digit to the right
+      val regex    = s"0{0,$newPos}(\\d+)".r
+      val stripped = newStr match {
+        case regex(remainder) => remainder
+        case _                => newStr
+      }
+      val s        = s"$minus$stripped"
+      (s, s, s.length - str.length)
+    } else {
+      (str, str, 0)
+    }
+  }
+
+  private def hasNDecimalsOrFewer(str: String, n: Int): Boolean = {
+    val decimalPos = str.indexOf(".")
+    decimalPos < 0 || decimalPos > str.length - n - 2
+  }
+
+  private def isValidNDigitInt(str: String, maxDigits: Int, maxValue: Int) =
+    if (str == "") true
+    else
+      str.length <= maxDigits && str.parseIntOption
+        .map(i => 0 <= i && i <= maxValue)
+        .getOrElse(false)
```

Comment:
You can shorten `.map(...).getOrElse(false)` to `.exists(...)`.

Reply:
doh!